api: improve listVM API handling and response speed

rohityadavcloud opened the pull request:
- Changes behaviour of details param handling for:
  - listVirtualMachines API: when the detail param is not provided, it uses `all` details except `stats`
  - listVirtualMachinesMetrics API: when the detail param is not provided, it uses `all` details including `stats`
- Remove ConfigKey vm.stats.increment.metrics.in.memory which was renamed to `vm.stats.increment.metrics` in #5984
- Changes default value of VM stats accumulation setting `vm.stats.increment.metrics` to false until a better solution emerges. Since #5984, this is true and during the execution of listVM APIs the stats are clubbed/calculated which can immensely slow down list VM API calls.
- Fix UI that uses listVirtualMachinesMetrics to not call `stats` detail when in list view without metrics selected.

These changes see 2-4x gains in the listVirtualMachines APIs call and in the UI. For environment where code changes are not possible, disabling `vm.stats.increment.metrics` global setting saw 2-4x speed gain in the list API calls.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [x] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial


### Screenshots (if appropriate):


### How Has This Been Tested?

Tested a workaround in a 4.18.2.0 env KVM based, with only 12 VMs:

1. List VM API call without details and with details but global setting `vm.stats.increment.metrics` disabled:

<img width="1097" alt="Screenshot 2024-04-26 at 11 29 31 AM" src="https://github.com/apache/cloudstack/assets/95203/67d134d9-71c6-4f9e-a159-18516786eb07">

2. List VM API call without detail and with detail and `vm.stats.increment.metrics` enabled:
<img width="1067" alt="Screenshot 2024-04-26 at 11 31 15 AM" src="https://github.com/apache/cloudstack/assets/95203/e8473829-e3c0-424a-ba46-7f863eae6272">

- We see ~3-4x gains by simply disabling the global setting
- We see 15x gains when stats details isn't used but the global setting is enabled
- We see 4x gains when stats detail isn't used and global setting is disabled

With this PR, I propose to change the earlier behaviour for listVirtualMachines API to exclude stats when details aren't provided. This can immensely speed up several integrations, who call listVMs APIs that not necessarily need `stats` in it.
While, the listVirtualMachinesMetrics API was amended to retain old behaviour so when details aren't passed it could
return all details including stats as expected. In the release notes likely we can mention about this, so folks who don't want accumulated stats can disable it but for new env we don't want to encourage this by default (as we already have the nice vm history metrics feature/tab).

#### How did you try to break this feature and the system with this change?

<!-- see how your change affects other areas of the code, etc. -->


<!-- Please read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/main/CONTRIBUTING.md) document -->

rohityadavcloud commented:
@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

weizhouapache reviewed (commented):
code looks ok

my only concern is the backward compatibility
we may need a highlighted section to list breaking changes

rohityadavcloud commented:
@weizhouapache this is breaking largely wrt 4.17 but not much as earlier versions; the only issue is for listVirtualMachines API which IMO shouldn't return vmstats (by looping through accumulating stats while the API is made). But I agree it could be an edge case where this break somebody's integration who aren't using the listVirtualMachinesMetrics instead (which doesn't break the compatibility).

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 9435

rohityadavcloud commented:
@blueorangutan test

JoaoJandre commented:
Hello @rohityadavcloud , @weizhouapache  

Looking briefly at this PR I noticed that it has a big intersection with #8782, especially the changes on the listVirtualMachines and listVirtualMachineMetrics APIs.

When #8782 proposed to change the default behavior of them to not list the stats, both @weizhouapache  and @rohityadavcloud  were firmly against changing the default behavior (see https://github.com/apache/cloudstack/pull/8782#issuecomment-1995297443 and https://github.com/apache/cloudstack/pull/8782#issuecomment-1997341885). So now, only a month later, we get this PR: what is the big difference on these two PRs? And if there is, why not just propose an adjustment to #8782? 

In any case, this PR not only breaks compatibility on the APIs, but also changes the default value of another configuration; and all this is going into 4.18! Are we going to introduce breaking changes on a maintenance version? I would really like to understand what made you change your opinion on compatibility so fast.

sureshanaparti reviewed (approved):
clgtm

blueorangutan commented:
<b>[SF] Trillian test result (tid-10044)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 43454 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8985-t10044-kvm-centos7.zip
Smoke tests completed. 109 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_04_nonsecured_to_secured_vm_migration | `Error` | 166.37 | test_vm_life_cycle.py
test_08_migrate_vm | `Error` | 46.21 | test_vm_life_cycle.py

rohityadavcloud commented:
Thanks @weizhouapache @JoaoJandre I've addressed the feedback and comments.

Joao's remarks are valid, however, my comments at the time were largely around metrics vs non-metrics APIs - how they should be used treated. At the time I couldn't co-relate what made the API slow only to later find out through community and customer issues who are facing degraded performance.

Generally speaking, we shouldn't break API compatibility which can break potential integrations (as also on this PR Wei has objected). I feel in this case whatever we do we break what was introduced by #5984  in 4.17. The root cause that I could determine is, we accumulate VM stats for each VM in list API calls which isn't a good idea. That said, I think we need a stop-gap until a better solution emerges on how we should process and return stats are returned by the list API (I think we shouldn't reduce/accumulate stats during a list API call).

The list VM API response method is used not only for listVMs API but also for several other VM related lifecycle APIs (such as migrate, stop/start etc) which all return the same vm response class/object. So, I've tried to have a different approach with the PR, and introduce a new global setting that can allow users who want backward compatibility but largely benefit users who don't want it (as default). I've updated the description to reflect this.

rohityadavcloud commented:
@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 9452

rohityadavcloud commented:
@blueorangutan test

blueorangutan commented:
@rohityadavcloud a [SL] Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-10065)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 45766 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8985-t10065-kvm-centos7.zip
Smoke tests completed. 128 look OK, 2 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_events_resource | `Error` | 417.66 | test_events_resource.py
test_01_redundant_vpc_site2site_vpn | `Failure` | 409.80 | test_vpc_vpn.py
test_01_vpc_site2site_vpn | `Failure` | 294.07 | test_vpc_vpn.py

rohityadavcloud commented:
@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 9461

GutoVeronezi commented:
> At the time I couldn't co-relate what made the API slow only to later find out through community and customer issues who are facing degraded performance.

That is the main point of the PR #8782 and I explained it in https://github.com/apache/cloudstack/pull/8737#issuecomment-1995191529, which is a discussion you participated, @rohityadavcloud. 

It is not the first time that I have seen some going against a PR and sometime later the same guys or group proposing the same thing. @JoaoJandre had already proposed the same type of solution on #8782 and received several comments against it; all lacking technical reason to be against it. And now we are facing a PR from those who were against João's PR doing the same thing. This is a duplicate of PR #8782; I am -1 on this PR and PR #8782 should be prioritized.

shwstppr reviewed (approved):
Code LGTM

@blueorangutan package

rohityadavcloud commented:
As I don't care much for the PR anymore and #8782 has addressed most if not all my concerns, I'm closing this PR. I may raise a new PR for changes that aren't in #8782 and we probably may need a doc PR in release notes.
